File: DDDSample.Domain/Aggregates/Customer/CustomerAggregate.vb
Imports DDDSample

Public Class CustomerAggregate
    Public ReadOnly Changes As IList(Of IEvent) = New List(Of IEvent)
    ReadOnly _state As CustomerState

    Public Sub New(events As IEnumerable(Of IEvent))
        _state = New CustomerState(events)
    End Sub
    Public Sub Create(id As CustomerId, name As String)
        If _state.Created Then Throw New InvalidOperationException("Customer Already Created")


    End Sub

    Public Sub Apply(e As IEvent)
        _state.Mutate(e)
        Changes.Add(e)
    End Sub
End Class

File: DDDSample.Domain/Aggregates/Customer/CustomerAggregate.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace DDDSample
{
     public class CustomerAggregate
    {
         public readonly IList<IEvent> Changes = new List<IEvent>();
         readonly CustomerState _state ;

         public CustomerAggregate(IEnumerable<IEvent> events)
         {
             _state = new CustomerState(events);
         }
        public void Create(CustomerId id, string name)
        {
             if (_state.Created) throw new InvalidOperationException("Customer Already Created");
             Apply(new CustomerCreated(id,name));
        }
       
        public void Apply(IEvent e)
        {
            _state.Mutate(e);
            Changes.Add(e);
        }

    }
}
